---
layout: page
title: Readings
permalink: /readings/
---

\* Denotes a recommended reading <br>
\*\* Denotes a highly recommended reading


### Week 1: Jan 10 - 14

[Volz, E. M., Koelle, K., & Bedford, T. (2013). Viral phylodynamics. PLoS Computational Biology, 9(3), e1002947.][volz2013] <br>
\* Provides a bird's-eye overview of phylodynamics <br>

[volz2013]: <https://doi.org/10.1371/journal.pcbi.1002947>

### Week 2: Jan 17 - 21

[Pereira, R., Oliveira, J., & Sousa, M. (2020). Bioinformatics and Computational Tools for Next-Generation Sequencing Analysis in Clinical Genetics. Journal of Clinical Medicine, 9(1), 132.][pereira2020] <br>

[pereira2020]: https://www.mdpi.com/2077-0383/9/1/132/htm

### Week 3: Jan 24 - 28

[Felsenstein , J. (1981). Evolutionary trees from DNA sequences: A maximum likelihood approach. J. Mol. Evol., 17(6) 368-376.][felsenstein1981] <br>
Introduces the Felsenstein pruning algorithm <br>

[felsenstein1981]: https://link.springer.com/article/10.1007/BF01734359

[Holder, M., & Lewis, P. O. (2003). Phylogeny estimation: traditional and Bayesian approaches. Nature Reviews Genetics, 4(4), 275.][holder2003] <br>
\* Reviews key aspects of Bayesian phylogenetic inference <br>

[holder2003]: <http://cgl.bioinfo.uqam.ca/bif7001/articles/BIF7001-Phylo-NatureReviewsGenetics4a.pdf>

[Yang, Z. (2014). Molecular evolution: a statistical approach. Oxford University Press.][yang2014] <br>
\* Chapter 1 gives a great overview of the substitution models used in molecular evolution. <br>

[yang2014]: <https://catalog.lib.ncsu.edu/catalog/DUKE006146449>

### Week 4: Jan 31 - Feb 4

[Lemey, P., Rambaut, A., Drummond, A. J., & Suchard, M. A. (2009). Bayesian phylogeography finds its roots. PLoS Computational Biology, 5(9), e1000520.][lemey2009] <br>

[lemey2009]: <https://doi.org/10.1371/journal.pcbi.1000520>

[Pybus, O. G., Suchard, M. A., Lemey, P., Bernardin, F. J., Rambaut, A., Crawford, F. W., ... & Delwart, E. L. (2012). Unifying the spatial epidemiology and molecular evolution of emerging epidemics. PNAS, 109(37), 15066-15071.][pybus2012] <br>

[pybus2012]: <https://doi.org/10.1073/pnas.1206598109>

[De Maio, N., Wu, C. H., O’Reilly, K. M., & Wilson, D. (2015). New routes to phylogeography: a Bayesian structured coalescent approximation. PLoS Genetics, 11(8)][demaio2015] <br>
\* Demonstrates how structured coalescent models can improve upon discrete trait phylogeographic analysis <br>

[demaio2015]: <https://doi.org/10.1371/journal.pgen.1005421>

### Week 5: Feb 7 - 11

[Rosenberg, N. A., & Nordborg, M. (2002). Genealogical trees, coalescent theory and the analysis of genetic polymorphisms. Nature Reviews Genetics, 3(5), 380-390.][rosenberg2002] <br>
\* Provides a great conceptual overview of coalescent theory <br>

[rosenberg2002]: <https://www.nature.com/articles/nrg795>

[Drummond, A. J., Rambaut, A., Shapiro, B. E. T. H., & Pybus, O. G. (2005). Bayesian coalescent inference of past population dynamics from molecular sequences. Molecular Biology and Evolution, 22(5), 1185-1192.][drummond2005] <br>

[drummond2005]: <https://academic.oup.com/mbe/article/22/5/1185/1066885>

### Week 6: Feb 14 - 18

[Jombart, T., Cori, A., Didelot, X., Cauchemez, S., Fraser, C., & Ferguson, N. (2014). Bayesian reconstruction of disease outbreaks by combining epidemiologic and genomic data. PLoS Computational Biology, 10(1)][jombart2014]

[jombart2014]: <https://journals.plos.org/ploscompbiol/article/file?type=printable&id=10.1371/journal.pcbi.1003457>

[De Maio, N., Wu, C. H., & Wilson, D. J. (2016). SCOTTI: efficient reconstruction of transmission within outbreaks with the structured coalescent. PLoS Computational Biology, 12(9).][demaio2016]

[demaio2016]: <https://doi.org/10.1371/journal.pcbi.1005130>

[Wymant, C., Hall, M., Ratmann, O., Bonsall, D., Golubchik, T., de Cesare, M., ... and The BEEHIVE Collaboration. (2018). PHYLOSCANNER: inferring transmission from within-and between-host pathogen genetic diversity. Molecular Biology and Evolution, 35(3), 719-733.][wymant2018]

[wymant2018]: <https://academic.oup.com/mbe/article/35/3/719/4653772>

### Week 7: Feb 21 - 25

[Hein, J., Schierup, M., & Wiuf, C. (2004). Gene genealogies, variation and evolution: a primer in coalescent theory. Oxford University Press, USA.][hein2004] <br>
\*\* Chapter 5 presents an excellent overview of recombination and its effect on phylogenies. <br>

[hein2004]: <https://catalog.lib.ncsu.edu/catalog/NCSU1940857>

[Boni, M. F., Posada, D., & Feldman, M. W. (2007). An exact nonparametric method for inferring mosaic structure in sequence triplets. Genetics, 176(2), 1035-1047.][boni2007] 

[boni2007]: <https://www.genetics.org/content/176/2/1035.short>

### Week 8: Feb 28 - March 4

[Shapiro, B. J. (2016). How clonal are bacteria over time?. Current Opinion in Microbiology, 31, 116-123.][shapiro2016]

[shapiro2016]: https://doi.org/10.1016/j.mib.2016.03.013

### Week 9: March 7 - 11

[Stadler, T., & Bonhoeffer, S. (2013). Uncovering epidemiological dynamics in heterogeneous host populations using phylogenetic methods. Philosophical Transactions of the Royal Society B: Biological Sciences, 368(1614), 20120198.][stadler2013] <br>
\* Describes the multi-type birth-death model for pathogen phylogenies. <br>

[stadler2013]: <https://doi.org/10.1098/rstb.2012.0198>

[Kühnert, D., Kouyos, R., Shirreff, G., Pečerska, J., Scherrer, A. U., Böni, J., ... & Stadler, T. (2018). Quantifying the fitness cost of HIV-1 drug resistance mutations through phylodynamics. PLoS Pathogens, 14(2), e1006895.][kuhnert2018] 

[kuhnert2018]: <https://doi.org/10.1371/journal.ppat.1006895>

### Week 10: March 21 - 25

[Keeling, M. J., & Rohani, P. (2011). Modeling infectious diseases in humans and animals. Princeton University Press.][keeling2011] <br>
\*\* Chapters 2 and 3 give an amazing introduction to SIR-type models. Unfortunately not available online but well worth it if you can get your hands on a copy. <br>

[keeling2011]: <https://catalog.lib.ncsu.edu/catalog/NCSU2255422>

[Ferrari, M. J., Grais, R. F., Bharti, N., Conlan, A. J., Bjørnstad, O. N., Wolfson, L. J., ... & Grenfell, B. T. (2008). The dynamics of measles in sub-Saharan Africa. Nature, 451(7179), 679-684.][ferrari2008] <br>

[ferrari2008]: <https://www.nature.com/articles/nature06509>

[Gilligan, C. A., & van den Bosch, F. (2008). Epidemiological models for invasion and persistence of pathogens. Annu. Rev. Phytopathol., 46, 385-418.][gilligan2008] <br>
\* Review exploring many different applications of epidemiological modeling to plant pathogens <br>

[gilligan2008]: <https://www.annualreviews.org/doi/abs/10.1146/annurev.phyto.45.062806.094357>

### Week 11: March 21 - 25

[Keeling, M. J., & Rohani, P. (2011). Modeling infectious diseases in humans and animals. Princeton University Press.][keeling2011] <br>
\*\* Chapter 6 gives a great overview of the types of stochastic models used in epidemiology. <br>

[keeling2011]: <https://catalog.lib.ncsu.edu/catalog/NCSU2255422>

[Vaughan, T. G., & Drummond, A. J. (2013). A stochastic simulator of birth–death master equations with application to phylodynamics. Molecular Biology and Evolution, 30(6), 1480-1493.][vaughan2013] <br>

[vaughan2013]: <https://doi.org/10.1093/molbev/mst057>

### Week 12: April 4 - 8

[Volz, E. M., Pond, S. L. K., Ward, M. J., Brown, A. J. L., & Frost, S. D. (2009). Phylodynamics of infectious disease epidemics. Genetics, 183(4), 1421-1430.][volz2009] <br>
\* This paper first derived a coalescent model for SIR-type epidemiological models. <br>

[volz2009]: <https://doi.org/10.1534/genetics.109.106021>

[Rasmussen, D. A., Boni, M. F., & Koelle, K. (2014). Reconciling phylodynamics with epidemiology: the case of dengue virus in southern Vietnam. Molecular Biology and Evolution, 31(2), 258-271.][rasmussen2014]

[rasmussen2014]: <https://doi.org/10.1093/molbev/mst203>

[Volz, E. M., & Siveroni, I. (2018). Bayesian phylodynamic inference with complex models. PLoS Computational Biology, 14(11), e1006546.][phydyn] <br>

[phydyn]: <https://doi.org/10.1371/journal.pcbi.1006546>

### Week 13: April 11 - 15

[Didelot, X., & Parkhill, J. (2021). A scalable analytical approach from bacterial genomes to epidemiology. bioRxiv][didelot2021]

[didelot2021]: https://doi.org/10.1101/2021.11.19.469232

### Week 14: April 18 - 22

[Łuksza, M., & Lässig, M. (2014). A predictive fitness model for influenza. Nature, 507(7490), 57-61.][luksza2014] <br>

[luksza2014]: <https://www.nature.com/articles/nature13087>

[Morris, D. H., Gostic, K. M., Pompei, S., Bedford, T., Łuksza, M., Neher, R. A., ... & McCauley, J. W. (2018). Predictive modeling of influenza shows the promise of applied evolutionary biology. Trends in Microbiology, 26(2), 102-118.][morris2018] <br>

[morris2018]: <https://doi.org/10.1016/j.tim.2017.09.004>

